### PR TITLE
feat: add inventory and other cli cmds

### PR DESCRIPTION
**Pull request by nadim-az**

### CODE REVIEW

File: ordersettler/ordersettler.go [hunk 1]
```diff
@@ -108,120 +105,39 @@ func (r *OrderSettler) Run(ctx context.Context) {
...
-			}
-
-			// ensure order is not already filled (an order is only marked as
-			// filled on the source chain once it is settled)
-			status, err := sourceBridgeClient.OrderStatus(ctx, sourceGatewayAddress, fill.OrderID)
-			if err != nil {
-				return fmt.Errorf("getting order %s status on chainID %s: %w", fill.OrderID, sourceChainID, err)
-			}
-			if status != fast_transfer_gateway.OrderStatusUnfilled {
-				r.ordersSeen[fill.OrderID] = true
```
Comment (dhfang) [inline ordersettler/ordersettler.go:188]: the order settlement discovery logic has been refactored into a separate function, but all the logic that is used to skip rpc calls for orders that have already been seen has been discarded. it should not require any general context about the order settler to understand that this function pre and post refactor are not logically equivalent.

File: cmd/solvercli/cmd/settlements.go [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+	amount *big.Int
+	count  int
+}
+
+var settlementsCmd = &cobra.Command{
+	Use:   "settlements",
+	Short: "Show pending settlement amounts across chains",
+	Run: func(cmd *cobra.Command, args []string) {
+		ctx := setupContext(cmd)
+		keysPath, err := cmd.Flags().GetString("keys")
```
Comment (dhfang) [inline cmd/solvercli/cmd/settlements.go:31]: one of the desired goals that was expressed for the settlement command is that we should be able to check the pending settlements for any arbitrary solver. can we do that if we require the private keys to be specified?

File: cmd/solvercli/cmd/balances.go [hunk 3]
```diff
@@ -0,0 +1,290 @@
...
+			Balance:    balance,
+			Symbol:     "USDC",
+			Decimals:   CCTP_TOKEN_DECIMALS,
+		}
+	}
+
+	return balances, gasBalances, nil
+}
+
+func getCustomAssetsBalances(ctx context.Context,
```
Comment (dhfang) [inline cmd/solvercli/cmd/balances.go:198]: why do we rely on two methods of fetching asset balances here, it seems like fetching usdc and gas balances can just rely on the balances endpoint?

**Review by dhfang — CHANGES_REQUESTED** [review]

File: cmd/solvercli/cmd/settlements.go (hunk 2 above)
Comment (nadim-az) [reply cmd/solvercli/cmd/settlements.go:31]: yes, this is for chandra
Screenshot 2024-12-05 at 2 16 23 AM

File: ordersettler/ordersettler.go (hunk 1 above)
Comment (nadim-az) [reply ordersettler/ordersettler.go:188]: my bad I accidentally put the ordersSeen check to after the rpc calls, fixed

File: cmd/solvercli/cmd/balances.go (hunk 3 above)
Comment (nadim-az) [reply cmd/solvercli/cmd/balances.go]: good point, updated

File: README.md [hunk 4]
```diff
@@ -111,6 +111,32 @@ make build-cli # build cli tool
 --checkpoint-storage-location-override <hplStorageFileOverride> \
 ```
 
+balances: show current on-chain balances for USDC, gas token, and any custom assets requested
+
+```shell
+./build/solvercli balances \
```
Comment (mattac21) [inline README.md:117]: now that the cli is getting a lot more functionality, i think we should provide instructions for users to build the cli and then add it to their path, then they can just run `solver balances ...` instead of `./build/solvercli balances ...` (and maybe rename from solvercli to just solver) (this could also be in another pr, but imo it will feel more like a polished cli then)

File: cmd/solvercli/cmd/balances.go [hunk 5]
```diff
@@ -0,0 +1,245 @@
...
+	ChainID           string
+	Balance           *big.Int
+	Symbol            string
+	Decimals          uint8
+	WarningThreshold  *big.Int
+	CriticalThreshold *big.Int
+}
+
+var balancesCmd = &cobra.Command{
+	Use:   "balances",
```
Comment (mattac21) [inline cmd/solvercli/cmd/balances.go:62]: we should document the `--custom-asssets` flag in an area besides just the example usage. it took me a big to figure out where that was coming from in the code. I think we can put it in the `Use` field https://github.com/spf13/cobra/blob/v1.8.1/command.go#L51, or maybe the `Args`

File: cmd/solvercli/cmd/balances.go [hunk 6]
```diff
@@ -0,0 +1,245 @@
...
+	Short: "Show current on-chain balances (USDC and gas tokens)",
+	Long: `Show current on-chain balances for USDC and gas tokens across all configured chains.
+    
+Example:
```
Comment (mattac21) [inline cmd/solvercli/cmd/balances.go:66]: the example should probably go in the `Example` field

File: cmd/solvercli/cmd/balances.go [hunk 7]
```diff
@@ -0,0 +1,245 @@
...
+		}
+	},
+}
+
+func init() {
+	rootCmd.AddCommand(balancesCmd)
+	balancesCmd.Flags().String("custom-assets", "", "JSON map of chain IDs to denom arrays, e.g. '{\"osmosis\":[\"uosmo\",\"uion\"],\"celestia\":[\"utia\"]}'")
+}
+
+func getBalances(ctx context.Context, cmd *cobra.Command) (map[string]*ChainBalance, map[string]*ChainGasBalance, map[string][]*ChainBalance, *big.Int, *big.Float, error) {
```
Comment (mattac21) [inline cmd/solvercli/cmd/balances.go:117]: this feels like it should be a couple different functions, from looking at it, maybe a function to gather up what chain + asset pairs the user wants to get the balances of, one to get the actual balances response, and one to parse? maybe some other split makes more sense, but usually when there are 3 or more returns the func should probably be broken up

File: cmd/solvercli/cmd/settlements.go (hunk 2 above)
Comment (mattac21) [reply cmd/solvercli/cmd/settlements.go:31]: i feel like we should add an option to not use a keys file and instead use a flag to specify the osmosis address and evm address. using the keys file is nice (and probably a good default for solvers) when checking your own balances, but for us when we want to check on other solvers balances, creating keys files for all of them just to check the inventory/settlements is going to get tiring)

Comment (mattac21) [reply cmd/solvercli/cmd/settlements.go:31]: im assuming you are requiring the keys file since the client manager needs it, but you could probably make a new constructor that allows a smaller set of functions (non signing functions) that doesnt require the keys to be specified

File: README.md [hunk 8]
```diff
@@ -111,6 +111,32 @@ make build-cli # build cli tool
...
+  --config ./config/local/config.yml \
+  --custom-assets '{"osmosis-1":["uosmo","uion"],"celestia-1":["utia"]}'
+```
+
+inventory: show complete solver inventory including balances, settlements, and rebalances
```
Comment (mattac21) [inline README.md:122]: should specify that the settlements and rebalances returned here are _pending_ settlements and rebalances

File: cmd/solvercli/cmd/balances.go (hunk 5 above)
Comment (mattac21) [reply cmd/solvercli/cmd/balances.go:62]: https://github.com/spf13/cobra/blob/v1.8.1/command.go#L51 actually this is probably the right way https://github.com/spf13/cobra/blob/main/site/content/user_guide.md#local-flags

File: cmd/solvercli/cmd/balances.go (hunk 6 above)
Comment (mattac21) [reply cmd/solvercli/cmd/balances.go:66]: same for other commands as well

File: README.md (hunk 4 above)
Comment (nadim-az) [reply README.md]: good shout, done

File: cmd/solvercli/cmd/balances.go (hunk 5 above)
Comment (nadim-az) [reply cmd/solvercli/cmd/balances.go]: yessir i've already added an explanation there

File: cmd/solvercli/cmd/balances.go (hunk 6 above)
Comment (nadim-az) [reply cmd/solvercli/cmd/balances.go]: done

File: cmd/solvercli/cmd/balances.go (hunk 7 above)
Comment (nadim-az) [reply cmd/solvercli/cmd/balances.go]: i refactored the logic a bit here, take a look and lmk if its easier to follow now :D

File: README.md (hunk 8 above)
Comment (nadim-az) [reply README.md]: done

File: cmd/solvercli/cmd/settlements.go (hunk 2 above)
Comment (nadim-az) [reply cmd/solvercli/cmd/settlements.go:33]: I ran this without a keys file setup specifically for chandra, just the latest default local/keys.json that is in main. the keystore values aren't really being used here, its the addresses configured in the config.yml file

i agree that it would be cleaner to have a separate client manager, can this be done in a separate PR though? this one is quite big already and this seems a lower priority task, since it doesn't affect us checking other solvers balances

I updated the balances command though to accept custom address flags, and default to the addresses in the config otherwise. Doing this for settlements is a bigger refactor since we depend on the config directly in the ordersettler process, and would be out of scope for this PR if we want to do that. The current experience for us to check other solvers balances is just needing to update the config solver_address field, which isnt too painful imo

File: cmd/solvercli/cmd/balances.go [hunk 9]
```diff
@@ -0,0 +1,331 @@
...
+			address = evmAddress
+		} else if chainConfig.Type == config.ChainType_COSMOS && osmosisAddress != "" {
+			address = osmosisAddress
+		}
+
+		var gasTokenDenom string
+		if chainConfig.Type == config.ChainType_COSMOS {
+			gasTokenDenom = chainConfig.Cosmos.GasDenom
+		} else if chainConfig.Type == config.ChainType_EVM {
+			gasTokenDenom = chainConfig.ChainName + "-native"
```
Comment (nadim-az) [inline cmd/solvercli/cmd/balances.go:195]: this is a janky solution, but our balances endpoint doesn't accept ETH as a denom, and returns a balance but no decimals/value_usd for a 0x000.. address. This wasn't an issue previously because we were using rpc to get the gas balance.

Two alternatives are: updating our balances endpoint to accept eth parameter or handle the 0x000 address more gracefully (by setting the usd_value and decimals field in return), or we could alternatively add an eth_representation field in our config but I didnt want to add more noise in our config since there's alot there already @dhfang @mattac21

Comment (mattac21) [reply cmd/solvercli/cmd/balances.go:195]: did you end up getting this sorted by passing `ethereum-native` as the denom? or `arbitrum-native`, `optimism-native`, etc?

Comment (nadim-az) [reply cmd/solvercli/cmd/balances.go:195]: arbitrum-native, etc.

ChainName + "-native"

File: Makefile [hunk 10]
```diff
@@ -24,6 +24,7 @@ build: ${SERVER_BIN}
 
 .PHONY: build-cli
 build-cli: ${CLI_BIN}
+	sudo sh -c 'cp ${CLI_BIN} /usr/local/bin/solver && chmod 755 /usr/local/bin/solver'
```
Comment (mattac21) [inline Makefile:27]: maybe this is just me but I feel like this probablyyyy shouldn't be using sudo under the hood. imo we should just build the cli and then maybe print out like hey you will have to run these commands to move the solvercli binary into your path.

File: README.md [hunk 11]
```diff
@@ -79,36 +79,64 @@ make db-clean # clean all existing db entries
 
 ### CLI Tool
 
-submit-order: submit a fast transfer order to the deployed gateway contract specifying an amount in USDC and a recipient.
-This only supports transferring USDC from EVM -> Osmosis, as any pre/post transfer swaps are typically handled by the [Skip Go](https://docs.skip.build/go/general/getting-started) API.
+First, build and install the CLI from the project root directory:
 
 ```shell
 make build-cli
```
Comment (mattac21) [inline README.md:85]: maybe below this we could have the commands you need to move the solver into the users PATH

File: cmd/solvercli/cmd/balances.go [hunk 12]
```diff
@@ -0,0 +1,331 @@
...
+	balancesCmd.Flags().String("osmosis-address", "", "Optional Osmosis address to check balances for instead of config address")
+}
+
+func getBalances(ctx context.Context,
+	usdcBalances map[string]*ChainBalance,
+	gasBalances map[string]*ChainGasBalance,
+	customBalances map[string][]*ChainBalance,
+	totalUSDCBalance *big.Int,
+	totalCustomAssetsUSDValue *big.Float,
+	cmd *cobra.Command) error {
```
Comment (mattac21) [inline cmd/solvercli/cmd/balances.go:128]: nit, should be this for consistency 
```suggestion
func getBalances(
        ctx context.Context,
	usdcBalances map[string]*ChainBalance,
	gasBalances map[string]*ChainGasBalance,
	customBalances map[string][]*ChainBalance,
	totalUSDCBalance *big.Int,
	totalCustomAssetsUSDValue *big.Float,
	cmd *cobra.Command
) error {
```

File: cmd/solvercli/cmd/balances.go (hunk 9 above)
Comment (mattac21) [reply cmd/solvercli/cmd/balances.go:195]: oh duh i see what you're saying now, tbh i think its fine as is. maybe the logic could be moved out to a function on its own, but its probably fine

**Review by mattac21 — DISMISSED** [review]

**Review by mattac21 — APPROVED** [review]

File: cmd/solvercli/cmd/settlements.go [hunk 13]
```diff
@@ -0,0 +1,94 @@
...
+			lmt.Logger(ctx).Fatal("Error reading key-store-type", zap.Error(err))
+		}
+
+		keyStore, err := keys.GetKeyStore(keyStoreType, keys.GetKeyStoreOpts{KeyFilePath: keysPath})
+		if err != nil {
+			lmt.Logger(ctx).Fatal("Unable to load keystore", zap.Error(err))
+		}
+
+		cosmosTxExecutor := cosmos.DefaultSerializedCosmosTxExecutor()
+		cctpClientManager := clientmanager.NewClientManager(keyStore, cosmosTxExecutor)
```
Comment (dhfang) [inline cmd/solvercli/cmd/settlements.go:49]: seems like we can just pass a nil value for the key store if it's irrelevant to the function of this command, or does that create an error?

File: ordersettler/detector.go [hunk 14]
```diff
@@ -0,0 +1,112 @@
...
+		}
+
+		for _, fill := range fills {
+			if ordersSeen != nil && ordersSeen[fill.OrderID] {
+				continue
+			}
+
+			sourceChainID, err := config.GetConfigReader(ctx).GetChainIDByHyperlaneDomain(strconv.Itoa(int(fill.SourceDomain)))
+			if err != nil {
+				continue
```
Comment (dhfang) [inline ordersettler/detector.go:60]: the existing code has logic here that marks an order as seen. please go and double check these functions are logically equivalent pre and post refactor

File: cmd/solvercli/cmd/balances.go [hunk 15]
```diff
@@ -0,0 +1,333 @@
...
+	Decimals          uint8
+	WarningThreshold  *big.Int
+	CriticalThreshold *big.Int
+}
+
+var balancesCmd = &cobra.Command{
+	Use:     "balances",
+	Short:   "Show current on-chain balances for USDC, gas tokens, and custom assets across all configured chains.",
+	Long:    `Show current on-chain balances for USDC, gas tokens, and custom assets across all configured chains.`,
+	Example: `solver balances --custom-assets '{"osmosis-1":["uosmo","uion"],"celestia-1":["utia"]}'`,
```
Comment (dhfang) [inline cmd/solvercli/cmd/balances.go:65]: I would have assumed the balances command would be a generic asset checking command without this distinction between usdc+gas tokens and everything else. The inventory command would just invoke it specifically with usdc and the gas tokens. Will leave it to your discretion whether or not you change that.

**Review by dhfang — CHANGES_REQUESTED** [review]

File: ordersettler/detector.go [hunk 16]
```diff
@@ -51,12 +48,19 @@ func DetectPendingSettlements(
...
 
 			sourceChainID, err := config.GetConfigReader(ctx).GetChainIDByHyperlaneDomain(strconv.Itoa(int(fill.SourceDomain)))
 			if err != nil {
+				lmt.Logger(ctx).Warn(
+					"failed to get source chain ID by hyperlane domain. skipping order settlement. it may be unsettled.",
+					zap.Uint32("hyperlaneDomain", fill.SourceDomain),
+					zap.String("orderID", fill.OrderID),
+					zap.Error(err),
+				)
+				(*ordersSeen)[fill.OrderID] = true
```
Comment (dhfang) [inline ordersettler/detector.go:63]: why did we make the map a pointer? a map can already be nil. you also have the potential to dereference a nil pointer here.

**Review by dhfang — CHANGES_REQUESTED** [review]

File: ordersettler/detector.go (hunk 16 above)
Comment (nadim-az) [reply ordersettler/detector.go]: sorry i didnt realize i pushed this change, this is obviously wrong

the refactor i ultimately wanted to do was for the detector to return a newOrdersSeen map, that the ordersettler can then use to update its own internal state. but not super necessary right now

**Review by dhfang — APPROVED** [review]
